readthedocs/doc_builder/environments.py

```diff
@@ -612,128 +547,31 @@ def run_command_class(self, *cmd, **kwargs):  # pylint: disable=arguments-differ
         })
         return super().run_command_class(*cmd, **kwargs)
 
-    @property
-    def successful(self):
-        """Build completed, without top level failures or failing commands."""
-        return (
-            self.done and self.failure is None and
-            all(cmd.successful for cmd in self.commands)
-        )
-
-    @property
-    def failed(self):
-        """Is build completed, but has top level failure or failing commands."""
-        return (
-            self.done and (
-                self.failure is not None or
-                any(cmd.failed for cmd in self.commands)
-            )
-        )
-
-    @property
-    def done(self):
-        """Is build in finished state."""
-        return (
-            self.build and
-            self.build['state'] == BUILD_STATE_FINISHED
-        )
-
-    def update_build(self, state=None):
```

Comment:
Yes, this code was moved to the task itself.

The general rule behind this was: "The task should be in charge of everything related to updating the build/version/project. On the other hand, the environment should only be in charge of running the commands + hitting the API with the command's output because it's the only one that has access to that data"

Reply:
By moving this logic to the task, it's _always_ clear _when_ we need to update the build and _how_ depending on where we are in the flow (e.g. `after_return` the build is FINISHED; while installing dependencies, the build is INSTALLING, etc). The logic got simplified a lot.

Besides, we always update the build in the same way: update the `self.build` dictionary and call `self.update_build` from the task. This also may reduce the number of hits to the API required (e.g. we could update the `self.build` dictionary _without_ hitting the API immediately if required).

This is also useful together with `on_success`, `on_failure` and `after_return` where the first two methods update the `self.build['success']` with a specific value, and then in `after_return` we hit the API _without knowing_ if the build was successful or not.

Reply:
Some of the tasks and environment changes I still want to spend some time with and digest, but so far the changes to the pattern here make sense to me, and I get how environment interacts with the build task at a high level.